In [1]:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

import pandas as pd

import warnings
warnings.filterwarnings('ignore')

In [2]:
trainOri = pd.read_csv('./input/train.csv')
testOri = pd.read_csv('./input/test.csv')

train = trainOri.copy()
test = testOri.copy()

In [3]:
train.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [4]:
test.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities      ...       ScreenPorch PoolArea PoolQC  Fence  \
0         Lvl    AllPub      ...               120        0    NaN  MnPrv   
1         Lvl    AllPub      ...                 0        0    NaN    NaN   
2         Lvl    AllPub      ...                 0        0    NaN  MnPrv   
3         Lvl    AllPub      ...                 0        0    NaN    NaN   
4         HLS    AllPub      ...               144        0    NaN    NaN   

  MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0         NaN       0      6    2010        WD         Normal  
1        Gar2   12500      6    2010        WD         Normal  
2         NaN       0      3    2010        WD         Normal  
3         NaN       0      6    2010        WD         Normal  
4         NaN       0      1    2010        WD         Normal  

[5 rows x 80 columns]

获取存在缺失值的特征,同时计算缺失率

In [5]:
aa = train.isnull().sum()
aa[aa>0].sort_values(ascending=False)/len(train)

PoolQC          0.995205
MiscFeature     0.963014
Alley           0.937671
Fence           0.807534
FireplaceQu     0.472603
LotFrontage     0.177397
GarageYrBlt     0.055479
GarageType      0.055479
GarageFinish    0.055479
GarageQual      0.055479
GarageCond      0.055479
BsmtFinType2    0.026027
BsmtExposure    0.026027
BsmtFinType1    0.025342
BsmtCond        0.025342
BsmtQual        0.025342
MasVnrArea      0.005479
MasVnrType      0.005479
Electrical      0.000685
dtype: float64

In [6]:
aa = test.isnull().sum()
aa[aa>0].sort_values(ascending=False)/len(test)

PoolQC          0.997944
MiscFeature     0.965045
Alley           0.926662
Fence           0.801234
FireplaceQu     0.500343
LotFrontage     0.155586
GarageYrBlt     0.053461
GarageCond      0.053461
GarageQual      0.053461
GarageFinish    0.053461
GarageType      0.052090
BsmtCond        0.030843
BsmtExposure    0.030158
BsmtQual        0.030158
BsmtFinType1    0.028787
BsmtFinType2    0.028787
MasVnrType      0.010966
MasVnrArea      0.010281
MSZoning        0.002742
BsmtFullBath    0.001371
BsmtHalfBath    0.001371
Utilities       0.001371
Functional      0.001371
Exterior2nd     0.000685
Exterior1st     0.000685
SaleType        0.000685
BsmtFinSF1      0.000685
BsmtFinSF2      0.000685
BsmtUnfSF       0.000685
KitchenQual     0.000685
GarageCars      0.000685
GarageArea      0.000685
TotalBsmtSF     0.000685
dtype: float64

In [7]:
textFeaturesTr = train.select_dtypes(include=[object])
print(len(textFeaturesTr.columns))
print(textFeaturesTr.columns)

43
Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition'],
      dtype='object')


In [8]:
textFeaturesTe = test.select_dtypes(include=[object])
print(len(textFeaturesTe.columns))
print(textFeaturesTe.columns)

43
Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition'],
      dtype='object')


对特征分组：

MSSubClass、MSZoning

LotFrontage

LotArea

Street、Alley

LotShape、LandContour

Utilities

LotConfig、LandSlope、Neighborhood、Condition1、Condition2

BldgType、HouseStyle

OverallQual、OverallCond

YearBuilt、YearRemodAdd

RoofStyle、RoofMatl、Exterior1st、Exterior2nd、MasVnrType、MasVnrArea、ExterQual、ExterCond、Foundation

BsmtQual、BsmtCond、BsmtExposure、BsmtFinType1、BsmtFinSF1、BsmtFinType2、BsmtFinSF2、BsmtUnfSF、TotalBsmtSF

Heating、HeatingQC、CentralAir、Electrical

1stFlrSF、2ndFlrSF、LowQualFinSF、GrLivArea

BsmtFullBath、BsmtHalfBath、FullBath、HalfBath、Bedroom、Kitchen、KitchenQual、TotRmsAbvGrd

Functional

Fireplaces、FireplaceQu

GarageType、GarageYrBlt、GarageFinish、GarageCars、GarageArea、GarageQual、GarageCond

PavedDrive、WoodDeckSF

OpenPorchSF、EnclosedPorch、3SsnPorch、ScreenPorch

PoolArea、PoolQC

Fence

MiscFeature

MoSold、YrSold

SaleType、SaleCondition

In [9]:
train['MSSubClass'].value_counts()

20     536
60     299
50     144
120     87
30      69
160     63
70      60
80      58
90      52
190     30
85      20
75      16
45      12
180     10
40       4
Name: MSSubClass, dtype: int64

In [10]:
def missing_rate(df,idx):
    print(df[idx].isnull().sum()/len(df))

## Prepare the Data

### Data Cleaning

In [11]:
#有缺失值的字段
aa = train.isnull().sum()
aa[aa>0].sort_values(ascending=False)/len(train)

PoolQC          0.995205
MiscFeature     0.963014
Alley           0.937671
Fence           0.807534
FireplaceQu     0.472603
LotFrontage     0.177397
GarageYrBlt     0.055479
GarageType      0.055479
GarageFinish    0.055479
GarageQual      0.055479
GarageCond      0.055479
BsmtFinType2    0.026027
BsmtExposure    0.026027
BsmtFinType1    0.025342
BsmtCond        0.025342
BsmtQual        0.025342
MasVnrArea      0.005479
MasVnrType      0.005479
Electrical      0.000685
dtype: float64

In [12]:
aa = train.isnull().sum()
missingColsTr= aa[aa>0]
missingColsTr = missingColsTr.index.tolist()
print(len(missingColsTr))
print(missingColsTr)

19
['LotFrontage', 'Alley', 'MasVnrType', 'MasVnrArea', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageQual', 'GarageCond', 'PoolQC', 'Fence', 'MiscFeature']


一共19个特征出现空缺值。

将填充方式相同的特征集合到一起，一块填充

In [13]:
aa = test.isnull().sum()
missingColsTe= aa[aa>0]
missingColsTe = missingColsTe.index.tolist()
print(len(missingColsTe))
print(missingColsTe)

33
['MSZoning', 'LotFrontage', 'Alley', 'Utilities', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'BsmtFullBath', 'BsmtHalfBath', 'KitchenQual', 'Functional', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType']


In [14]:
def show_categorical(df, cols):
    print(df[cols].value_counts())

In [15]:
for col in missingCols:
    show_categorical(train,col)
    print("="*64)

NameError: name 'missingCols' is not defined

In [16]:
#使用None填充：string类型
cols = ['Alley','MasVnrType','BsmtQual','BsmtCond','BsmtExposure','BsmtFinType1','BsmtFinType2','Electrical',\
       'FireplaceQu','GarageType','GarageFinish','GarageQual','GarageCond','PoolQC','Fence','MiscFeature']
colsZero = ['MasVnrArea','GarageYrBlt']
#'GarageYrBlt'
#'LotFrontage'
for col in cols:
    train[col].fillna('NA',inplace=True)

for col in colsZero:
    train[col].fillna(0,inplace=True)

In [17]:
#LotFrontage
train[~train.LotFrontage.isnull()].head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave    NA      Reg   
1   2          20       RL         80.0     9600   Pave    NA      Reg   
2   3          60       RL         68.0    11250   Pave    NA      IR1   
3   4          70       RL         60.0     9550   Pave    NA      IR1   
4   5          60       RL         84.0    14260   Pave    NA      IR1   

  LandContour Utilities    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0     NA    NA          NA       0   
1         Lvl    AllPub    ...            0     NA    NA          NA       0   
2         Lvl    AllPub    ...            0     NA    NA          NA       0   
3         Lvl    AllPub    ...            0     NA    NA          NA       0   
4         Lvl    AllPub    ...            0     NA    NA          NA       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [18]:
train['LotFrontage'].fillna(train['LotFrontage'].median(),inplace=True)
#用中位数填充

In [19]:
train[train.LotFrontage.isnull()].head()

Empty DataFrame
Columns: [Id, MSSubClass, MSZoning, LotFrontage, LotArea, Street, Alley, LotShape, LandContour, Utilities, LotConfig, LandSlope, Neighborhood, Condition1, Condition2, BldgType, HouseStyle, OverallQual, OverallCond, YearBuilt, YearRemodAdd, RoofStyle, RoofMatl, Exterior1st, Exterior2nd, MasVnrType, MasVnrArea, ExterQual, ExterCond, Foundation, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinSF1, BsmtFinType2, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, Heating, HeatingQC, CentralAir, Electrical, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, KitchenQual, TotRmsAbvGrd, Functional, Fireplaces, FireplaceQu, GarageType, GarageYrBlt, GarageFinish, GarageCars, GarageArea, GarageQual, GarageCond, PavedDrive, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, PoolQC, Fence, MiscFeature, MiscVal, MoSold, YrSold, SaleType, SaleCondition, SalePrice]
Index: []

[0 rows x 81 columns]

In [20]:
# 测试集填充
show_categorical(test, missingColsTe)

AttributeError: 'DataFrame' object has no attribute 'value_counts'

**value_counts适用于离散型变量的值分布统计**

In [21]:
print(len(missingColsTe))
print(missingColsTe)

33
['MSZoning', 'LotFrontage', 'Alley', 'Utilities', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'BsmtFullBath', 'BsmtHalfBath', 'KitchenQual', 'Functional', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType']


In [22]:
#使用None填充：string类型
cols = ['Alley','MasVnrType','BsmtQual','BsmtCond','BsmtExposure','BsmtFinType1','BsmtFinType2','Electrical',\
       'FireplaceQu','GarageType','GarageFinish','GarageQual','GarageCond','PoolQC','Fence','MiscFeature']
colsZero = ['BsmtFinSF1','BsmtFinSF2','BsmtUnfSF','TotalBsmtSF','BsmtFullBath',\
            'BsmtHalfBath','GarageYrBlt','GarageCars','GarageArea']
# 众数 MSZoning Utilities Exterior1st Exterior2nd KitchenQual Functional SaleType
colsMode = ['MSZoning','Utilities','Exterior1st','Exterior2nd','KitchenQual','Functional','SaleType']
# 中位数 'LotFrontage' MasVnrArea
colsMedian = ['LotFrontage','MasVnrArea']
for col in cols:
    test[col].fillna('NA',inplace=True)

for col in colsZero:
    test[col].fillna(0,inplace=True)

for col in colsMode:
    test[col].fillna(test[col].mode()[0],inplace=True)
    
for col in colsMedian:
    test[col].fillna(test[col].median(),inplace=True)


In [23]:
aa = test.isnull().sum()
aa[aa>0].sort_values(ascending=False)

Series([], dtype: int64)

发现用众数mode填充失败！为什么？
Series 众数返回是一个Series
添加一个下标就行了



数据填充完毕；

特征划分，类别型、数值型

分别进行处理

In [24]:
Ids = test['Id']
test = test.drop('Id', axis=1)

targets = train['SalePrice']
train = train.drop(['Id','SalePrice'],axis=1)

In [25]:
print(train.shape)
print(test.shape)

(1460, 79)
(1459, 79)


In [26]:
#将年份等相关特征从int64 转变成 str类型
def chg(df, cols):
    for col in cols:
        df[col] = df[col].astype('str')

cols = ['YearBuilt','YearRemodAdd','YrSold','MoSold','MSSubClass','GarageYrBlt']

chg(train,cols)
chg(test,cols)

In [27]:
cateFeats = ['ExterQual', 'ExterCond','HeatingQC', 'KitchenQual', 'FireplaceQu',\
             'GarageQual', 'GarageCond','BsmtQual','BsmtCond','BsmtFinType1',\
             'BsmtFinType2','BsmtExposure','PoolQC','GarageType','GarageFinish',\
             'Fence','Functional','MiscFeature']

oneHotFeats = ['MSSubClass', 'MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour',\
               'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1',\
               'Condition2', 'BldgType', 'HouseStyle', 'YearBuilt', 'YearRemodAdd',\
               'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',\
               'Foundation', 'Heating', 'CentralAir', 'Electrical', 'PavedDrive',\
               'MoSold', 'YrSold', 'SaleType', 'SaleCondition','GarageYrBlt']
numFeats = train.select_dtypes(include=[np.int64,np.float64]).columns

In [28]:
print(len(cateFeats)+len(oneHotFeats)+len(numFeats))

79


In [29]:
train.shape

(1460, 79)

In [30]:
numFeats

Index(['LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'MasVnrArea',
       'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF',
       '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath',
       'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF',
       'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal'],
      dtype='object')

特征数量相等。

**数据集特征字段依据类型划分完成，之后根据特征类型采用不用的处理方法完成数据处理，处理之后，整合形成新的数据集；最后进行模型训练。**

### 类别编码

有顺序关系的特征编码

In [31]:
cateTrain = train[cateFeats]
cateTest = test[cateFeats]
cateTrain.head()

ExterQual ExterCond HeatingQC KitchenQual FireplaceQu GarageQual GarageCond  \
0        Gd        TA        Ex          Gd          NA         TA         TA   
1        TA        TA        Ex          TA          TA         TA         TA   
2        Gd        TA        Ex          Gd          TA         TA         TA   
3        TA        TA        Gd          Gd          Gd         TA         TA   
4        Gd        TA        Ex          Gd          TA         TA         TA   

  BsmtQual BsmtCond BsmtFinType1 BsmtFinType2 BsmtExposure PoolQC GarageType  \
0       Gd       TA          GLQ          Unf           No     NA     Attchd   
1       Gd       TA          ALQ          Unf           Gd     NA     Attchd   
2       Gd       TA          GLQ          Unf           Mn     NA     Attchd   
3       TA       Gd          ALQ          Unf           No     NA     Detchd   
4       Gd       TA          GLQ          Unf           Av     NA     Attchd   

  GarageFinish Fence Functional MiscFeature  
0          RFn    NA        Typ          NA  
1          RFn    NA        Typ          NA  
2          RFn    NA        Typ          NA  
3          Unf    NA        Typ          NA  
4          RFn    NA        Typ          NA

对类别数据one-hot编码

In [32]:
qual5toCat = {'Ex':5,'Gd':4,'TA':3,'Fa':2,'Po':1,'NA':0} # ExterQual ExterCond HeatingQC KitchenQual
# 6 FireplaceQu  GarageQual GarageCond
qual6toCatBsmt = {'NA':1,'Po':2,'Fa':3,'TA':4,'Gd':5,'Ex':6} # BsmtQual BsmtCond
qual7toCat = {'NA':1,'Unf':2,'LwQ':3,'Rec':4,'BLQ':5,'ALQ':6,'GLQ':7}# BsmtFinType1 BsmtFinType2

mapBsmtExposure = {'NA':1,'No':2,'Mn':3,'Av':4,'Gd':5} # BsmtExposure
mapPoolQC = {'Ex':5,'Gd':4,'TA':3,'Fa':2,'NA':1}#PoolQC
mapFunctional = {'Sal':1,'Sev':2,'Maj2':3,'Maj1':4,'Mod':5,'Min2':6,'Min1':7,'Typ':8}
mapGarageType = {'NA':1,'Detchd':2,'CarPort':3,'BuiltIn':4,'Basment':5,'Attchd':6,'2Types':7}
mapGarageFinish = {'NA':1,'Unf':2,'RFn':3,'Fin':4}
mapFence = {'NA':1,'MnWw':2,'GdWo':3,'MnPrv':4,'GdPrv':5}
mapMiscFeature = {'NA':1,'TenC':2,'Shed':3,'Othr':4,'Gar2':5,'Elev':6}

feats1 = ['ExterQual','ExterCond','HeatingQC','KitchenQual','FireplaceQu','GarageQual','GarageCond']
feats2 = ['BsmtQual','BsmtCond']
feats3 = ['BsmtFinType1','BsmtFinType2']
def transformDF(df):
    for feature in feats1:
        df = df.replace({feature: qual5toCat})
    for feature in feats2:
        df = df.replace({feature: qual6toCatBsmt})
    for feature in feats3:
        df = df.replace({feature: qual7toCat})
    df = df.replace({'BsmtExposure':mapBsmtExposure})
    df = df.replace({'PoolQC':mapPoolQC})
    df = df.replace({'Functional':mapFunctional})
    df = df.replace({'GarageType':mapGarageType})
    df = df.replace({'GarageFinish':mapGarageFinish})
    df = df.replace({'Fence':mapFence})
    df = df.replace({'MiscFeature':mapMiscFeature})
    return df

cateTrain = transformDF(cateTrain)
cateTest = transformDF(cateTest)

In [33]:
aa = cateTrain.isnull().sum()
aa[aa>0]

Series([], dtype: int64)

In [34]:
aa = cateTest.isnull().sum()
aa[aa>0]

Series([], dtype: int64)

In [35]:
cateTrain.head()

ExterQual  ExterCond  HeatingQC  KitchenQual  FireplaceQu  GarageQual  \
0          4          3          5            4            0           3   
1          3          3          5            3            3           3   
2          4          3          5            4            3           3   
3          3          3          4            4            4           3   
4          4          3          5            4            3           3   

   GarageCond  BsmtQual  BsmtCond  BsmtFinType1  BsmtFinType2  BsmtExposure  \
0           3         5         4             7             2             2   
1           3         5         4             6             2             5   
2           3         5         4             7             2             3   
3           3         4         5             6             2             2   
4           3         5         4             7             2             4   

   PoolQC  GarageType  GarageFinish  Fence  Functional  MiscFeature  
0       1           6             3      1           8            1  
1       1           6             3      1           8            1  
2       1           6             3      1           8            1  
3       1           2             2      1           8            1  
4       1           6             3      1           8            1

In [36]:
cateTest.head()

ExterQual  ExterCond  HeatingQC  KitchenQual  FireplaceQu  GarageQual  \
0          3          3          3            3            0           3   
1          3          3          3            4            0           3   
2          3          3          4            3            3           3   
3          3          3          5            4            4           3   
4          4          3          5            4            0           3   

   GarageCond  BsmtQual  BsmtCond  BsmtFinType1  BsmtFinType2  BsmtExposure  \
0           3         4         4             4             3             2   
1           3         4         4             6             2             2   
2           3         5         4             7             2             2   
3           3         4         4             7             2             2   
4           3         5         4             6             2             2   

   PoolQC  GarageType  GarageFinish  Fence  Functional  MiscFeature  
0       1           6             2      4           8            1  
1       1           6             2      1           8            5  
2       1           6             4      4           8            1  
3       1           6             4      1           8            1  
4       1           6             3      1           8            1

**次序类别特征处理完成**

### 没有顺序关系的特征
用one-hot encoder编码



In [37]:
oneHotTrain = train[oneHotFeats]
oneHotTest = test[oneHotFeats]

In [38]:
oneHotTrain.MSSubClass.value_counts()

20     536
60     299
50     144
120     87
30      69
160     63
70      60
80      58
90      52
190     30
85      20
75      16
45      12
180     10
40       4
Name: MSSubClass, dtype: int64

In [39]:
oneHotTest.MSSubClass.value_counts()

20     543
60     276
50     143
120     95
30      70
70      68
160     65
80      60
90      57
190     31
85      28
75       7
180      7
45       6
40       2
150      1
Name: MSSubClass, dtype: int64

In [40]:
print(type(oneHotTrain))
print(type(oneHotTest))

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>


In [41]:
oneHotTrain.Utilities.value_counts()

AllPub    1459
NoSeWa       1
Name: Utilities, dtype: int64

In [42]:
oneHotTest.Utilities.value_counts()

AllPub    1459
Name: Utilities, dtype: int64

**转换后数据长度不相同，主要是由于特征值的缺失；有的值在训练集特征中出现，在测试集中却没有出现，或者反之。**

In [43]:
# #现将MSSubClass的数字转换成类别码A～Z
# oneHotTrain = oneHotTrain.replace({'MSSubClass':{20: "A", 30: "B", 40: "C", 45: "D", 50: "E",\
#                                      60: "F", 70: "G", 75: "H", 80: "I", 85: "J",\
#                                      90: "K", 120: "L", 150: "M", 160: "N", 180: "O", 190: "P"}})

# oneHotTest = oneHotTest.replace({'MSSubClass':{20: "A", 30: "B", 40: "C", 45: "D", 50: "E",\
#                                      60: "F", 70: "G", 75: "H", 80: "I", 85: "J",\
#                                      90: "K", 120: "L", 150: "M", 160: "N", 180: "O", 190: "P"}})

In [44]:
# 对MSSubClass单独进行one-hot编码
dummyMSSubClassTrain = oneHotTrain['MSSubClass']
dummyMSSubClassTest = oneHotTest['MSSubClass']

oneHotTrain = oneHotTrain.drop('MSSubClass',axis=1)
oneHotTest = oneHotTest.drop('MSSubClass',axis=1)

In [45]:
print(oneHotTrain.shape)
print(oneHotTest.shape)

(1460, 30)
(1459, 30)


In [46]:
dummyMSSubClassTrain = pd.get_dummies(dummyMSSubClassTrain,prefix='MSSubClass')

In [47]:
dummyMSSubClassTrain.head()

MSSubClass_120  MSSubClass_160  MSSubClass_180  MSSubClass_190  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   MSSubClass_20  MSSubClass_30  MSSubClass_40  MSSubClass_45  MSSubClass_50  \
0              0              0              0              0              0   
1              1              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   MSSubClass_60  MSSubClass_70  MSSubClass_75  MSSubClass_80  MSSubClass_85  \
0              1              0              0              0              0   
1              0              0              0              0              0   
2              1              0              0              0              0   
3              0              1              0              0              0   
4              1              0              0              0              0   

   MSSubClass_90  
0              0  
1              0  
2              0  
3              0  
4              0

In [48]:
dummyMSSubClassTest = pd.get_dummies(dummyMSSubClassTest,prefix='MSSubClass')
dummyMSSubClassTest.head()

MSSubClass_120  MSSubClass_150  MSSubClass_160  MSSubClass_180  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               1               0               0               0   

   MSSubClass_190  MSSubClass_20  MSSubClass_30  MSSubClass_40  MSSubClass_45  \
0               0              1              0              0              0   
1               0              1              0              0              0   
2               0              0              0              0              0   
3               0              0              0              0              0   
4               0              0              0              0              0   

   MSSubClass_50  MSSubClass_60  MSSubClass_70  MSSubClass_75  MSSubClass_80  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              1              0              0              0   
3              0              1              0              0              0   
4              0              0              0              0              0   

   MSSubClass_85  MSSubClass_90  
0              0              0  
1              0              0  
2              0              0  
3              0              0  
4              0              0

In [49]:
print(dummyMSSubClassTrain.shape)
print(dummyMSSubClassTest.shape)

(1460, 15)
(1459, 16)


In [50]:
dummyMSSubClassTrain.columns

Index(['MSSubClass_120', 'MSSubClass_160', 'MSSubClass_180', 'MSSubClass_190',
       'MSSubClass_20', 'MSSubClass_30', 'MSSubClass_40', 'MSSubClass_45',
       'MSSubClass_50', 'MSSubClass_60', 'MSSubClass_70', 'MSSubClass_75',
       'MSSubClass_80', 'MSSubClass_85', 'MSSubClass_90'],
      dtype='object')

In [51]:
dummyMSSubClassTest.columns

Index(['MSSubClass_120', 'MSSubClass_150', 'MSSubClass_160', 'MSSubClass_180',
       'MSSubClass_190', 'MSSubClass_20', 'MSSubClass_30', 'MSSubClass_40',
       'MSSubClass_45', 'MSSubClass_50', 'MSSubClass_60', 'MSSubClass_70',
       'MSSubClass_75', 'MSSubClass_80', 'MSSubClass_85', 'MSSubClass_90'],
      dtype='object')

In [52]:
#解决两个shape不相同的情况
add150 = np.zeros((1460,1))
pd150 = pd.DataFrame(add150,columns=['MSSubClass_150'])
dummyMSSubClassTrain = pd.concat([dummyMSSubClassTrain, pd150],axis=1)

In [53]:
dummyMSSubClassTrain.columns

Index(['MSSubClass_120', 'MSSubClass_160', 'MSSubClass_180', 'MSSubClass_190',
       'MSSubClass_20', 'MSSubClass_30', 'MSSubClass_40', 'MSSubClass_45',
       'MSSubClass_50', 'MSSubClass_60', 'MSSubClass_70', 'MSSubClass_75',
       'MSSubClass_80', 'MSSubClass_85', 'MSSubClass_90', 'MSSubClass_150'],
      dtype='object')

In [54]:
dummyMSSubClassTest.columns

Index(['MSSubClass_120', 'MSSubClass_150', 'MSSubClass_160', 'MSSubClass_180',
       'MSSubClass_190', 'MSSubClass_20', 'MSSubClass_30', 'MSSubClass_40',
       'MSSubClass_45', 'MSSubClass_50', 'MSSubClass_60', 'MSSubClass_70',
       'MSSubClass_75', 'MSSubClass_80', 'MSSubClass_85', 'MSSubClass_90'],
      dtype='object')

In [55]:
dummyMSSubClassTrain.sort_index(axis=1,ascending = True,inplace=True)
dummyMSSubClassTrain.columns

Index(['MSSubClass_120', 'MSSubClass_150', 'MSSubClass_160', 'MSSubClass_180',
       'MSSubClass_190', 'MSSubClass_20', 'MSSubClass_30', 'MSSubClass_40',
       'MSSubClass_45', 'MSSubClass_50', 'MSSubClass_60', 'MSSubClass_70',
       'MSSubClass_75', 'MSSubClass_80', 'MSSubClass_85', 'MSSubClass_90'],
      dtype='object')

In [56]:
dummyMSSubClassTest.columns

Index(['MSSubClass_120', 'MSSubClass_150', 'MSSubClass_160', 'MSSubClass_180',
       'MSSubClass_190', 'MSSubClass_20', 'MSSubClass_30', 'MSSubClass_40',
       'MSSubClass_45', 'MSSubClass_50', 'MSSubClass_60', 'MSSubClass_70',
       'MSSubClass_75', 'MSSubClass_80', 'MSSubClass_85', 'MSSubClass_90'],
      dtype='object')

In [57]:
from sklearn.preprocessing import OneHotEncoder

encoder = OneHotEncoder(handle_unknown='ignore')#处理未知特征：ignore忽略；默认error 报错；

oneHotTrain = encoder.fit_transform(oneHotTrain.values)
# 稀疏矩阵转换成稠密型矩阵
oneHotTrain = oneHotTrain.toarray()

In [58]:
oneHotTrain.shape

(1460, 459)

In [59]:
oneHotTest = encoder.transform(oneHotTest.values)
oneHotTest = oneHotTest.toarray()

*MSSubClass 150: 训练集中没有，测试集中有。所以先对特征做个转换映射到新特征上，之后再用onehot encoder 学习，转换*

确实多一个150标签

OneHotEncoder将所有特征都进行转换了；不可取！

将各种类型数据切分，最后再整合到一起

In [60]:
print('oneHotTrain shape:', oneHotTrain.shape)
print('oneHotTest shape:', oneHotTest.shape)

oneHotTrain shape: (1460, 459)
oneHotTest shape: (1459, 459)


In [61]:
encoder.get_feature_names()

array(['x0_C (all)', 'x0_FV', 'x0_RH', 'x0_RL', 'x0_RM', 'x1_Grvl',
       'x1_Pave', 'x2_Grvl', 'x2_NA', 'x2_Pave', 'x3_IR1', 'x3_IR2',
       'x3_IR3', 'x3_Reg', 'x4_Bnk', 'x4_HLS', 'x4_Low', 'x4_Lvl',
       'x5_AllPub', 'x5_NoSeWa', 'x6_Corner', 'x6_CulDSac', 'x6_FR2',
       'x6_FR3', 'x6_Inside', 'x7_Gtl', 'x7_Mod', 'x7_Sev', 'x8_Blmngtn',
       'x8_Blueste', 'x8_BrDale', 'x8_BrkSide', 'x8_ClearCr',
       'x8_CollgCr', 'x8_Crawfor', 'x8_Edwards', 'x8_Gilbert',
       'x8_IDOTRR', 'x8_MeadowV', 'x8_Mitchel', 'x8_NAmes', 'x8_NPkVill',
       'x8_NWAmes', 'x8_NoRidge', 'x8_NridgHt', 'x8_OldTown', 'x8_SWISU',
       'x8_Sawyer', 'x8_SawyerW', 'x8_Somerst', 'x8_StoneBr', 'x8_Timber',
       'x8_Veenker', 'x9_Artery', 'x9_Feedr', 'x9_Norm', 'x9_PosA',
       'x9_PosN', 'x9_RRAe', 'x9_RRAn', 'x9_RRNe', 'x9_RRNn',
       'x10_Artery', 'x10_Feedr', 'x10_Norm', 'x10_PosA', 'x10_PosN',
       'x10_RRAe', 'x10_RRAn', 'x10_RRNn', 'x11_1Fam', 'x11_2fmCon',
       'x11_Duplex', 'x11_Twnhs', '

In [62]:
encoder.get_feature_names().tolist()

['x0_C (all)',
 'x0_FV',
 'x0_RH',
 'x0_RL',
 'x0_RM',
 'x1_Grvl',
 'x1_Pave',
 'x2_Grvl',
 'x2_NA',
 'x2_Pave',
 'x3_IR1',
 'x3_IR2',
 'x3_IR3',
 'x3_Reg',
 'x4_Bnk',
 'x4_HLS',
 'x4_Low',
 'x4_Lvl',
 'x5_AllPub',
 'x5_NoSeWa',
 'x6_Corner',
 'x6_CulDSac',
 'x6_FR2',
 'x6_FR3',
 'x6_Inside',
 'x7_Gtl',
 'x7_Mod',
 'x7_Sev',
 'x8_Blmngtn',
 'x8_Blueste',
 'x8_BrDale',
 'x8_BrkSide',
 'x8_ClearCr',
 'x8_CollgCr',
 'x8_Crawfor',
 'x8_Edwards',
 'x8_Gilbert',
 'x8_IDOTRR',
 'x8_MeadowV',
 'x8_Mitchel',
 'x8_NAmes',
 'x8_NPkVill',
 'x8_NWAmes',
 'x8_NoRidge',
 'x8_NridgHt',
 'x8_OldTown',
 'x8_SWISU',
 'x8_Sawyer',
 'x8_SawyerW',
 'x8_Somerst',
 'x8_StoneBr',
 'x8_Timber',
 'x8_Veenker',
 'x9_Artery',
 'x9_Feedr',
 'x9_Norm',
 'x9_PosA',
 'x9_PosN',
 'x9_RRAe',
 'x9_RRAn',
 'x9_RRNe',
 'x9_RRNn',
 'x10_Artery',
 'x10_Feedr',
 'x10_Norm',
 'x10_PosA',
 'x10_PosN',
 'x10_RRAe',
 'x10_RRAn',
 'x10_RRNn',
 'x11_1Fam',
 'x11_2fmCon',
 'x11_Duplex',
 'x11_Twnhs',
 'x11_TwnhsE',
 'x12_1.5Fin',
 '

In [63]:
oneHotTrain = pd.DataFrame(oneHotTrain,columns=encoder.get_feature_names().tolist())
oneHotTest = pd.DataFrame(oneHotTest,columns=encoder.get_feature_names().tolist())

In [64]:
oneHotTrain.head()

x0_C (all)  x0_FV  x0_RH  x0_RL  x0_RM  x1_Grvl  x1_Pave  x2_Grvl  x2_NA  \
0         0.0    0.0    0.0    1.0    0.0      0.0      1.0      0.0    1.0   
1         0.0    0.0    0.0    1.0    0.0      0.0      1.0      0.0    1.0   
2         0.0    0.0    0.0    1.0    0.0      0.0      1.0      0.0    1.0   
3         0.0    0.0    0.0    1.0    0.0      0.0      1.0      0.0    1.0   
4         0.0    0.0    0.0    1.0    0.0      0.0      1.0      0.0    1.0   

   x2_Pave     ...      x29_2001.0  x29_2002.0  x29_2003.0  x29_2004.0  \
0      0.0     ...             0.0         0.0         1.0         0.0   
1      0.0     ...             0.0         0.0         0.0         0.0   
2      0.0     ...             1.0         0.0         0.0         0.0   
3      0.0     ...             0.0         0.0         0.0         0.0   
4      0.0     ...             0.0         0.0         0.0         0.0   

   x29_2005.0  x29_2006.0  x29_2007.0  x29_2008.0  x29_2009.0  x29_2010.0  
0         0.0         0.0         0.0         0.0         0.0         0.0  
1         0.0         0.0         0.0         0.0         0.0         0.0  
2         0.0         0.0         0.0         0.0         0.0         0.0  
3         0.0         0.0         0.0         0.0         0.0         0.0  
4         0.0         0.0         0.0         0.0         0.0         0.0  

[5 rows x 459 columns]

In [65]:
oneHotTest.head()

x0_C (all)  x0_FV  x0_RH  x0_RL  x0_RM  x1_Grvl  x1_Pave  x2_Grvl  x2_NA  \
0         0.0    0.0    1.0    0.0    0.0      0.0      1.0      0.0    1.0   
1         0.0    0.0    0.0    1.0    0.0      0.0      1.0      0.0    1.0   
2         0.0    0.0    0.0    1.0    0.0      0.0      1.0      0.0    1.0   
3         0.0    0.0    0.0    1.0    0.0      0.0      1.0      0.0    1.0   
4         0.0    0.0    0.0    1.0    0.0      0.0      1.0      0.0    1.0   

   x2_Pave     ...      x29_2001.0  x29_2002.0  x29_2003.0  x29_2004.0  \
0      0.0     ...             0.0         0.0         0.0         0.0   
1      0.0     ...             0.0         0.0         0.0         0.0   
2      0.0     ...             0.0         0.0         0.0         0.0   
3      0.0     ...             0.0         0.0         0.0         0.0   
4      0.0     ...             0.0         0.0         0.0         0.0   

   x29_2005.0  x29_2006.0  x29_2007.0  x29_2008.0  x29_2009.0  x29_2010.0  
0         0.0         0.0         0.0         0.0         0.0         0.0  
1         0.0         0.0         0.0         0.0         0.0         0.0  
2         0.0         0.0         0.0         0.0         0.0         0.0  
3         0.0         0.0         0.0         0.0         0.0         0.0  
4         0.0         0.0         0.0         0.0         0.0         0.0  

[5 rows x 459 columns]

### 数值型数据

rescale
StandardScaler 正太化

In [66]:
numFeats

Index(['LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'MasVnrArea',
       'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF',
       '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath',
       'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF',
       'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal'],
      dtype='object')

In [67]:
numTrain = train[numFeats]
numTest = test[numFeats]

numTrain.head()

LotFrontage  LotArea  OverallQual  OverallCond  MasVnrArea  BsmtFinSF1  \
0         65.0     8450            7            5       196.0         706   
1         80.0     9600            6            8         0.0         978   
2         68.0    11250            7            5       162.0         486   
3         60.0     9550            7            5         0.0         216   
4         84.0    14260            8            5       350.0         655   

   BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  1stFlrSF   ...     Fireplaces  \
0           0        150          856       856   ...              0   
1           0        284         1262      1262   ...              1   
2           0        434          920       920   ...              1   
3           0        540          756       961   ...              1   
4           0        490         1145      1145   ...              1   

   GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  \
0           2         548           0           61              0          0   
1           2         460         298            0              0          0   
2           2         608           0           42              0          0   
3           3         642           0           35            272          0   
4           3         836         192           84              0          0   

   ScreenPorch  PoolArea  MiscVal  
0            0         0        0  
1            0         0        0  
2            0         0        0  
3            0         0        0  
4            0         0        0  

[5 rows x 30 columns]

In [68]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
numTrain = scaler.fit_transform(numTrain)
numTest = scaler.transform(numTest)

In [69]:
numTrain.shape

(1460, 30)

In [70]:
numTest.shape

(1459, 30)

In [71]:
numTrain

array([[-0.22087509, -0.20714171,  0.65147924, ..., -0.27020835,
        -0.06869175, -0.08768781],
       [ 0.46031974, -0.09188637, -0.07183611, ..., -0.27020835,
        -0.06869175, -0.08768781],
       [-0.08463612,  0.07347998,  0.65147924, ..., -0.27020835,
        -0.06869175, -0.08768781],
       ...,
       [-0.1754621 , -0.14781027,  0.65147924, ..., -0.27020835,
        -0.06869175,  4.95311151],
       [-0.08463612, -0.08016039, -0.79515147, ..., -0.27020835,
        -0.06869175, -0.08768781],
       [ 0.23325479, -0.05811155, -0.79515147, ..., -0.27020835,
        -0.06869175, -0.08768781]])

In [72]:
numTrain = pd.DataFrame(numTrain,columns=numFeats)

In [73]:
numTest = pd.DataFrame(numTest,columns=numFeats)

In [74]:
print(type(cateTrain),type(dummyMSSubClassTrain),type(oneHotTrain),type(numTrain))

<class 'pandas.core.frame.DataFrame'> <class 'pandas.core.frame.DataFrame'> <class 'pandas.core.frame.DataFrame'> <class 'pandas.core.frame.DataFrame'>


In [75]:
X_train = pd.concat([cateTrain,dummyMSSubClassTrain,oneHotTrain,numTrain],axis=1)
X_test = pd.concat([cateTest,dummyMSSubClassTest, oneHotTest,numTest],axis=1)

In [76]:
X_train.shape

(1460, 523)

In [77]:
X_test.shape

(1459, 523)

In [78]:
finalData = X_test.copy()

### PCA

对数据进行PCA降维

In [79]:
from sklearn.decomposition import PCA

pca = PCA(n_components=447)#0.90

X_scaled = pca.fit_transform(X_train)

In [80]:
X_test = pca.transform(X_test)

In [81]:
print(X_scaled.shape)
print(X_test.shape)

(1460, 447)
(1459, 447)


**对标签target对log处理**

In [164]:
y_log = np.log(targets)
#y_log1p = np.log1p(targets)# 加1处理：防止数字过小

不管特征数量，直接使用模型进行训练;

划分训练集和测试集:使用交叉验证

In [145]:
# from sklearn.model_selection import train_test_split

# X_train,X_test,y_train,y_test = train_test_split(X_train,targets,test_size=0.2,random_state=42)

In [146]:
from sklearn.model_selection import cross_val_score
def rmse_cv(model,X,y):
    rmse = np.sqrt(-cross_val_score(model, X, y, scoring="neg_mean_squared_error", cv=5))
    return rmse

In [151]:
from sklearn.linear_model import LinearRegression,Ridge,Lasso,SGDRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR,LinearSVR

linReg = LinearRegression()
ridge = Ridge()
lasso = Lasso()
sgdr = SGDRegressor()
rfr = RandomForestRegressor()
svr = SVR()
linsvr = LinearSVR()

names = ["LR", "Ridge", "Lasso", "RF", "SGD","SVR", "LinSVR"]
models = [linReg,ridge,lasso,rfr,sgdr,svr,linsvr]

for name, model in zip(names, models):
    score = rmse_cv(model, X_scaled, y_log)
    print("{}: {:.6f}, {:.4f}".format(name,score.mean(),score.std()))

LR: 0.272893, 0.0797
Ridge: 0.156057, 0.0241
Lasso: 0.352715, 0.0191
RF: 0.175900, 0.0077
SGD: 0.235319, 0.0251
SVR: 0.123422, 0.0120
LinSVR: 0.173017, 0.0368


没有对target做log处理：

LR: 64263.732879, 28575.9500
Ridge: 36544.265443, 7013.9259
Lasso: 46869.051568, 6440.3253
RF: 32961.069183, 7174.4391
SVR: 36431.104610, 4916.0113
LinSVR: 81134.153083, 6155.6159
SGD: 64263.732879, 28575.9500

In [148]:
linsvr.fit(X_scaled,y_log)

LinearSVR(C=1.0, dual=True, epsilon=0.0, fit_intercept=True,
     intercept_scaling=1.0, loss='epsilon_insensitive', max_iter=1000,
     random_state=None, tol=0.0001, verbose=0)

In [87]:
preds = linsvr.predict(X_test)

In [96]:
df = pd.DataFrame({'Id':Ids,'SalePrice':np.exp(preds)})
df.to_csv('./output/easy.csv',index=False)

**PCA之后，kaggle评测rmse从0.41869降低到0.18033；排名上升了565.**

**对标签target做log处理之后，rmse从0.18033变为0.15115，排名提升了538!** COOL!

### 尝试Ensembling

最简单的Stacking

**Simplest Stacking Strategy: Average the base models' scores!**

In [152]:
from sklearn.base import BaseEstimator,RegressorMixin,TransformerMixin,clone

In [153]:
# 新定义类Average scores

class AveragingModels(BaseEstimator, RegressorMixin, TransformerMixin):
    def __init__(self, models):
        self.models = models
        
    # 将原始模型clone一份，对克隆后模型进行训练、预测----保证不会影响到原始模型；毕竟这是一个求loss的函数，对原始数据影响越小越好
    def fit(self, X, y):
        self.models_ = [clone(x) for x in self.models]
        
        # 训练cloned模型
        for model in self.models_:
            model.fit(X, y)

        return self
    
    # 对每个模型在数据集上做预测，然后对各个模型的得分结果取平均
    def predict(self, X):
        predictions = np.column_stack([#将1D数组按照列横向拼接---列变多了，要保证每个1D数组长度相同
            model.predict(X) for model in self.models_
        ])
        return np.mean(predictions, axis=1)    

In [157]:
averaged_models = AveragingModels(models = (ridge,rfr,svr,linsvr))
score = rmse_cv(averaged_models,X_scaled,y_log)
print(" Averaged base models score: {:.4f} ({:.4f})\n".format(score.mean(), score.std()))

 Averaged base models score: 0.1356 (0.0182)



log1p处理: Averaged base models score: 0.1373 (0.0188)

log处理： Averaged base models score: 0.1356 (0.0182)

**Your submission scored 0.13095, which is an improvement of your previous score of 0.14981. Great job!**

 Averaged base models score: 0.1575 (0.0161)

In [107]:
averaged_models.fit(X_scaled,y_log)
pred_stack = averaged_models.predict(X_test)

In [108]:
df = pd.DataFrame({'Id':Ids,'SalePrice':np.exp(pred_stack)})
df.to_csv('./output/easyStackAverage.csv',index=False)

In [109]:
ss = rmse_cv(linsvr,X_scaled,y_log)
print(" Averaged base models score: {:.4f} ({:.4f})\n".format(ss.mean(), ss.std()))

 Averaged base models score: 0.1713 (0.0333)



**对标签target做log处理之后，rmse从0.15115变成了0.14981，排名提升了44! COOL!**

**To Be Continued......**

### **使用新的Ensemble方法：Stacking averaged Models Class**

![](http://i.imgur.com/QBuDOjs.jpg)

In [165]:
from sklearn.model_selection import KFold

In [181]:
class StackingAveragedModels(BaseEstimator,RegressorMixin,TransformerMixin):
    def __init__(self,base_models,meta_model,n_folds=5):
        self.base_models = base_models
        self.meta_model = meta_model
        self.n_folds = n_folds
    
    # 在clone模型上进行模型训练：保证不对原始模型造成影响；原始模型只是指定了一个模型类；
    def fit(self,X,y):
        self.base_models_ = [list() for x in self.base_models]#保存每个模型类下的n_folds个训练模型
        self.meta_model_ = clone(self.meta_model)
        Kfold = KFold(n_splits=self.n_folds,shuffle=True,random_state=42)
        
        #定义新特征保存矩阵：每个模型在n_folds-1个fold上训练，创建out-of-fold的预测结果，n_folds迭代结果后，out-of-fold预测结果
        #遍历所有训练集，形成新特征；最后n个模型的新特征在meta_model上训练即可；
        out_of_fold_predicts = np.zeros((X.shape[0], len(self.base_models)))
        
        for i, model in enumerate(self.base_models):#遍历所有模型，新特征维度 等于 模型的个数
            for train_index, holdout_index in Kfold.split(X,y):#对每个base模型进行训练，形成新特征
                instance = clone(model)
                self.base_models_[i].append(instance)
#                 print(self.base_models_[i])
                instance.fit(X[train_index],y.as_matrix()[train_index])
                y_pred = instance.predict(X[holdout_index])
                #第k次模型训练在留出折上的预测结果；n_folds遍历结果后，形成模型i的在训练集上的所有预测
                out_of_fold_predicts[holdout_index, i] = y_pred
        #在新特征上对meta_model进行训练
        self.meta_model_.fit(out_of_fold_predicts,y)
        
        return self
    
    #在对数据进行预测之前需要先将数据X转换过到新特征维度上，才能开始预测
    def predict(self,X):#axis=1 cross; axis=0 down
        meta_features = np.column_stack([
            np.column_stack([model.predict(X) for model in base_models]).mean(axis=1)#对每行取平均；最后剩一列
            for base_models in self.base_models_ ])
        return self.meta_model_.predict(meta_features)

In [182]:
np.any(np.isnan(X_scaled))

False

In [183]:
np.any(np.isnan(y_log))

False

In [184]:
print(type(X_scaled))
print(type(y_log))

<class 'numpy.ndarray'>
<class 'pandas.core.series.Series'>


将y_log从series转换成numpy形式

In [187]:
stacked_models = StackingAveragedModels(base_models=(ridge,sgdr,rfr,svr),meta_model=linReg)
score = rmse_cv(stacked_models,X_scaled,y_log)
print(" Averaged base models score: {:.4f} ({:.4f})\n".format(score.mean(), score.std()))

 Averaged base models score: 0.1265 (0.0145)



错误：ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

**原因是:送到模型的数据或者标签不是numpy形式，是dataframe或series导致的**

**解决方法：将数据转换成numpy**

In [188]:
stacked_models.fit(X_scaled, y_log)
stacked_train_pred = stacked_models.predict(X_scaled)
stacked_pred = stacked_models.predict(X_test)

0.10248793327764291


In [189]:
df = pd.DataFrame({'Id':Ids,'SalePrice':np.exp(stacked_pred)})
df.to_csv('./output/easyStack.csv',index=False)

**Your submission scored 0.12506, which is an improvement of your previous score of 0.13095. Great job!**

有帮助！！！COOL！！！